File: domain-driven-design/go-web/internal/application/services/order_service.go
// 应用层（协调领域逻辑，处理业务用例）：订单应用服务
package services

import (
  "fmt"
  "go-web-order/internal/domain/order"
  "math/rand"
  "sync"
  "time"
)

var (
  rng    = rand.New(rand.NewSource(time.Now().UnixNano()))
  rngMux sync.Mutex
)

// GenerateOrderID 生成时间戳 + 随机数的唯一订单号 (int)
func GenerateOrderID() int64 {
  rngMux.Lock()
  defer rngMux.Unlock()

  timestamp := time.Now().UnixMilli() // 毫秒时间戳
  random := rng.Intn(1000)            // 0-999 的随机数
  return int64(timestamp*1000) + int64(random)
}

// OrderService 订单应用服务，协调领域逻辑和业务用例
type OrderService struct {
  OrderRepository order.OrderRepository // 使用订单仓储接口
  // Repository order.OrderRepository // 使用通用接口声明
}

// 创建OrderService实例对象
func NewOrderService(orderRepo order.OrderRepository) *OrderService {
  return &OrderService{orderRepo}
}

// CreateOrder 创建订单并保存到仓储中
func (s *OrderService) CreateOrder(customerName string, amount float64) (*order.Order, error) {
  // 自动生成订单 ID ，实际应用中会采用分布式ID或者采用数据库自增键
  orderID := GenerateOrderID()
  newOrder := order.NewOrder(orderID, customerName, amount)
  if newOrder == nil {
    return nil, fmt.Errorf("订单创建失败")
  }

  // 保存订单
  err := s.OrderRepository.Save(newOrder)
  if err != nil {
    return nil, fmt.Errorf("订单保存失败: %v", err)
  }

  return newOrder, nil
}

// CancelOrder 取消订单
func (s *OrderService) CancelOrder(id int64) error {
  order, err := s.OrderRepository.FindByID(id) // 获取订单
  if err != nil {
    return fmt.Errorf("订单取消失败：%v", err)
  }

  order.Cancel()                      // 执行领域逻辑：取消订单
  err = s.OrderRepository.Save(order) // 保存更新后的订单
  if err != nil {
    return fmt.Errorf("订单取消失败：%v", err)
  }
  return nil
}

// GetOrder 查询订单
func (s *OrderService) GetOrder(id int64) (*order.Order, error) {
  return s.OrderRepository.FindByID(id)
}

// GetAllOrders 列出全部订单，此处省略分页
func (s *OrderService) GetAllOrders(userId int) ([]*order.Order, error) {
  return s.OrderRepository.FindAll(userId)
}

// UpdateOrder 更新订单的客户信息和金额
func (s *OrderService) UpdateOrder(id int64, customerName string, amount float64) (*order.Order, error) {
  // 获取订单
  order, err := s.OrderRepository.FindByID(id)
  if err != nil {
    return nil, fmt.Errorf("订单未找到: %v", err)
  }

  // 更新订单的客户信息和金额
  order.UpdateCustomerInfo(customerName)
  order.UpdateAmount(amount)

  // 保存更新后的订单
  err = s.OrderRepository.Save(order)
  if err != nil {
    return nil, fmt.Errorf("更新订单失败: %v", err)
  }

  return order, nil
}

// DeleteOrder 删除订单
func (s *OrderService) DeleteOrder(id int64) error {
  // 获取订单
  order, err := s.OrderRepository.FindByID(id)
  if err != nil {
    return fmt.Errorf("订单未找到: %v", err)
  }
  // 从仓储中删除订单
  err = s.OrderRepository.Delete(order.ID)
  if err != nil {
    return fmt.Errorf("删除订单失败: %v", err)
  }

  return nil
}

File: domain-driven-design/node-web/src/application/services/order-service.js
// 应用层（协调领域逻辑，处理业务用例）：订单应用服务
// src/application/services/order-service.js

import Order from '../../domain/order/order.js';

export default class OrderService {
  constructor(orderRepository) {
    this.orderRepository = orderRepository;
  }

  // 创建订单并保存到仓储中
  async createOrder(customerName, amount) {
    // 自动生成订单 ID
    const id = await this.generateOrderId();

    // 创建订单
    const newOrder = Order.newOrder(id, customerName, amount);
    if (!newOrder) {
      throw new Error('订单创建失败');
    }

    try {
      await this.orderRepository.save(newOrder);
      return newOrder;
    } catch (error) {
      throw new Error(`订单保存失败: ${error.message}`);
    }
  }

  // 自动生成订单 ID ，实际应用中会采用分布式ID或者采用数据库自增键
  async generateOrderId() {
    // 这里可以根据业务需求生成唯一的 ID。如：使用时间戳 + 随机数
    const timestamp = Date.now();
    const random = Math.floor(Math.random() * 1000);
    return parseInt(`${timestamp}${random}`);
  }

  // 取消订单
  async cancelOrder(id) {
    try {
      const order = await this.orderRepository.findByID(id);
      order.cancel();
      await this.orderRepository.save(order);
    } catch (error) {
      throw new Error(`订单取消失败: ${error.message}`);
    }
  }

  // 查询订单
  async getOrder(id) {
    try {
      return await this.orderRepository.findByID(id);
    } catch (error) {
      throw new Error(`查询订单失败: ${error.message}`);
    }
  }

  // 查询全部订单，省略分页
  async getAllOrders(userId) {
    try {
      return await this.orderRepository.findAll(userId);
    } catch (error) {
      throw new Error(`查询全部订单失败: ${error.message}`);
    }
  }

  // 更新订单的客户信息和金额
  async updateOrder(id, customerName, amount) {
    try {
      const order = await this.orderRepository.findByID(id);
      order.updateCustomerInfo(customerName);
      order.updateAmount(amount);
      await this.orderRepository.save(order);
      return order;
    } catch (error) {
      throw new Error(`更新订单失败: ${error.message}`);
    }
  }

  // 删除订单
  async deleteOrder(id) {
    try {
      const order = await this.orderRepository.findByID(id);
      await this.orderRepository.delete(order.id);
    } catch (error) {
      throw new Error(`删除订单失败: ${error.message}`);
    }
  }
}